pkg/clusterid/parse_cluster_id.go

```diff
@@ -5,13 +5,59 @@ import (
 
 	"github.com/pkg/errors"
 	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/buildinfo"
+	"github.com/stackrox/rox/pkg/logging"
 	"github.com/stackrox/rox/pkg/mtls"
+	"github.com/stackrox/rox/pkg/sync"
 )
 
+var (
+	once     sync.Once
+	instance *parserWrapper
```

Comment:
Is there a reason for using the wrapper? I was thinking about using the `Parser` instance here directly. 

All access to this should go through `GetParser()` anyway.

Reply:
Originally I had `instance` to be the `Parser` but there are a couple of reasons to wrap the actual implementation:
- `Override` is a member function of the wrapper, if we were to remove the wrapper we will have to make it a normal function adding possible data races when it comes to call `GetParser` and `Override`. I don’t think it is possible to have `Override` as a member function of the `parserImpl`.
- I also like the idea of not changing the singleton pointer. This is more of a philosophical preference though 🙂

Reply:
> if we were to remove the wrapper we will have to make it a normal function adding possible data races when it comes to call GetParser and Override.

I think this is the case with both approaches. You're exposing the wrapper (via `GetParser`) which can be changed via (internal pointer swapped) using `Override`. A caller that tries to use the wrapper via `GetParser` at the same time we have an override will cause a race. 

This test (when ran with `-race`) triggers a data race:
```go
func Test_Race(t *testing.T) {
	wrapper := GetParser()

	go func() {
		wrapper.Override(&testParser{})
	}()

	go func() {
		_, err := GetParser().parser.ParseClusterIDFromServiceCert(storage.ServiceType_BENCHMARK_SERVICE)
		assert.NoError(t, err)
	}()
}
```

```
==================
WARNING: DATA RACE
Read at 0x00c0003ab690 by goroutine 21:
  github.com/stackrox/rox/pkg/clusterid.Test_Race.func2()
      /Users/fbittenc/go/src/github.com/stackrox/stackrox/pkg/clusterid/parse_cluster_test.go:24 +0x37

Previous write at 0x00c0003ab690 by goroutine 20:
  github.com/stackrox/rox/pkg/clusterid.(*parserWrapper).Override()
      /Users/fbittenc/go/src/github.com/stackrox/stackrox/pkg/clusterid/override_parser_dev.go:15 +0x89
  github.com/stackrox/rox/pkg/clusterid.Test_Race.func1()
      /Users/fbittenc/go/src/github.com/stackrox/stackrox/pkg/clusterid/parse_cluster_test.go:20 +0x3e
```

Reply:
Ah! That makes sense. I've changed the implementation to get rid of the wrapper and added locks in `GetParser` and `Override`. I've tested this test with `-race` and didn't see any data races:

```golang
func TestWhatever(t *testing.T) {
	go func() {
		OverrideClusterIDParser(&parserImpl{})
	}()
	go func() {
		_, err := GetParser().ParseClusterIDFromServiceCert(storage.ServiceType_BENCHMARK_SERVICE)
		assert.NoError(t, err)
	}()
}
```